saleor/payment/utils.py

```diff
@@ -68,23 +73,18 @@ def create_payment(**payment_data):
     return payment
 
 
-def create_transaction(
-        payment: Payment,
-        token: str,
-        kind: str,
-        is_success: bool,
-        amount: Decimal,
-        currency: str,
-        gateway_response: Optional[Dict] = None) -> Transaction:
-    if not gateway_response:
+def create_transaction(payment: Payment, kind: str, payment_token: str,
+        gateway_response: Dict=None, **extra_params) -> Transaction:
+    """Creates a Transaction based on transaction kind and gateway response."""
+    if gateway_response is None:
         gateway_response = {}
     txn, _ = Transaction.objects.get_or_create(
         payment=payment,
-        token=token,
         kind=kind,
-        is_success=is_success,
-        amount=amount,
-        currency=currency,
+        token=gateway_response.get('transaction_id', payment_token),
+        is_success=gateway_response.get('is_success', False),
+        amount=extra_params.get('amount', payment.total),
```

Comment:
So you're saying even though we ask the gateway to say charge X amount, there's no certainty of that and the gateway can end up charging less/more? That's a bit inconvenient, but if that's the case then indeed the amount has to be provided in the response. Although we can't really validate that for all methods, can we? Not everything returns money, for example void doesn't work on amount, does it?

Reply:
Yes, it is very likely for some gateways, and the gateway will actually return success. So we can only trust the amount returned from the gateway. Othewise we may have discrepancy.

We can't make sure they are all validated in saleor's side but we should do that if we can. Some payment provider won't return the amount for some operations probably.

So in the gateway implementation, they should return ```amount``` from the raw response if they can. If they can't, they should still provide a default amount such as ```payment.total```.

Reply:
For the ```charge```, usually it is fine. It happens usually in ```refund``` or ```void```.  You ask for void X, but the payment provider will void less. It is because they don't support partial void or paitial refund.